drivers/tsl4531x/tsl4531x.c

```diff
@@ -0,0 +1,228 @@
+ * @author      Juan I Carrano <j.carrano@fu-berlin.de>
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * This driver was derived from the TSL2561 driver.
+ *
+ * @}
+ */
+
+#include <errno.h>
+
+#include "log.h"
+#include "tsl4531x.h"
+#include "tsl4531x_internals.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+#define _DATALOW 0
+#define _DATAHIGH 1
+
+int tsl4531x_init(tsl4531x_t *dev, const tsl4531x_params_t *params)
+{
+    int r;
+    uint8_t id;
+
+    /* Initialise I2C bus */
+    if ((r = i2c_acquire(params->i2c_dev)) < 0) {
+        DEBUG("I2C_dev is: %d.", params->i2c_dev);
+        DEBUG("[Error] Cannot acquire device. I2C error: %d\n", r);
+        return -ENODEV;
+    }
+
+    /* Test for connectivity - verify ID and compare against stored value */
+    if ((r = i2c_read_reg(params->i2c_dev,
+                          params->i2c_addr,
+                          TSL4531X_COMMAND(TSL4531X_ID_REG),
+                          &id, 0)) < 0) {
+        DEBUG("[Error] Cannot read ID register. I2C error: %d\n", r);
+        i2c_release(params->i2c_dev);
+        return -ENXIO;
+    }
+
+    DEBUG("[Info] tsl4531x sensor ID ? %d\n", id);
+
+    if (TSL4531X_GET_PARTNO(id) != params->part_number) {
+        DEBUG("[Error] not a TSL4531 sensor.\n");
+        i2c_release(params->i2c_dev);
+        return -ENOTSUP;
+    }
+
+    /* Configure device. In low power mode, we initially power the sensor down. */
+    if (((r = i2c_write_reg(params->i2c_dev,
+                            params->i2c_addr,
+                            TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                            TSL4531X_CONTROL(TSL4531X_POWER_MODE(params->low_power_mode)),
+                            0)) < 0)
+        ||
+        ((r = i2c_write_reg(params->i2c_dev,
+                            params->i2c_addr,
+                            TSL4531X_COMMAND(TSL4531X_CONFIGURATION_REG),
+                            TSL4531X_CONFIG(TSL4531X_PSAVESKIP_ON, params->integration_time),
+                            0)) < 0)) {
+
+        DEBUG("[Error] Cannot configure device. I2C error: %d\n", r);
+        i2c_release(params->i2c_dev);
+        return -ENXIO;
+    }
+
+    /* If device was configured correctly, initialise the device descriptor */
+    dev->i2c_dev = params->i2c_dev;
+    dev->i2c_addr = params->i2c_addr;
+    dev->integration_time = params->integration_time;
+    dev->low_power_mode = params->low_power_mode;
+    dev->high_power_mode_started_up = false;
+    dev->sample_start_time = 0; /* Device assumed to start up at same time as
+                                   microcontroller - i.e. when it hits this line */
+
+    i2c_release(params->i2c_dev);
+
+    return 0;
+}
+
+int tsl4531x_set_low_power_mode(tsl4531x_t *dev, bool low_power_mode)
+{
+    assert(dev);
+
+    int r;
+
+    dev->low_power_mode = low_power_mode;
+
+    if ((r = i2c_acquire(dev->i2c_dev) < 0)) {
```

Comment:
Thanks @codacy-bot for pointing this out.

Reply:
done